BrainPortal/lib/boutiques_input_value_fixer.rb

```diff
@@ -0,0 +1,192 @@
+#
+# In the descriptor, the spec would look like:
+#
+#    "custom": {
+#     "cbrain:integrator_modules": {
+#         "BoutiquesInputValueFixer": {
+#             "n_cpus": 1,
+#             "mem": "4G",
+#             "customquery": nil,
+#             "level": "group"
+#         }
+#     }
+# }
+# It is recommended to supply fixed_values in the specification is "closed" in the sense that
+# contains all the implied parameter fixes or deletions. The additional changes can be caused
+# either by pairwise dependencies between inputs or their values or
+#
+# If some of dependent/implied input
+# assignments or deletions are missed, the module will attempt to do add some implied chages automatically, yet
+# the present solution is crude, and does not guaranty best results.
+# Which, we believe, is still good enough for the major use case of this module : resource related params
+# (memory, threads, processes etc). Obviously, resource-related parameters
+# seldom have many dependecies or group constraints.
+#
+# *** Examples of implied parameter fixing ***
+#
+# Example 1 (Mild case) let parameters -min_mem_use and -max_mem_Gig and -all_mem be mutually exclusive. Then
+# fixing one option to be, say -max_mem_Gig to 16 without removing the other two
+# is not ideal. Indeed keeping options -max_mem_Gig and -all_mem
+# as user will be able to chose run tool with an option exclusive to already chosen -max_mem_Gig, resulting in
+# unpredicatable and likely erroneous results. Yet an experienced user, who might know well parameters of tool
+# still able to enter correct input values by ignoring -max_mem_Gig and -all_mem. This particular cases is addressed
+# by this module by deletion of unfixed elements of the group. However, generally speaking, the module
+# does not always guaranty that in the transformed descriptor
+# would allow parameter combinations, which are prevented by previous descriptor
+#
+# Example 2 (A more severe case of implied dependencies)
+# # change is possible. Let consider a case when the above boutiques has also 'one-is-required' flag for the same group.
+# In this case fixing value of one parameter while keeping one-is-required and flag, the
+#
+# Fixer aims at avoiding second case of the stopping user from entering inputs, possible with given choice of fixed inputs
+# yet does try hard at blocking all the combos impossible in the original boutiques.
+module BoutiquesInputValueFixer
+
+  # Note: to access the revision info of the module,
+  # you need to access the constant directly, the
+  # object method revision_info() won't work.
+  Revision_info=CbrainFileRevision[__FILE__] #:nodoc:
+
+  # the hash of param values to be fixed or be ommited ()  #
+  def fixed_values
+    self.boutiques_descriptor.custom_module_info('BoutiquesInputValueFixer')
+  end
+
+  require 'pry'
+
+  # deletes fixed inputs listed in the custom 'integrator_modules'
+  def descriptor_without_fixed_inputs(descriptor)
```

Comment:
Overall throughout this method you use `fixed_values.keys` a lot, and every time it is several layers of method calls that are executed just to get, over and over, the same set of input ids (something like `[ 'input1', 'input2', 'input3' ]`).

Can you just not, right at the beginning of the method, just extract this once?

Like in:

```ruby
fixed_input_ids = fixed_inputs.keys
```